src/emulator/auth/operations.ts

```diff
@@ -1838,6 +1811,47 @@ function validateCustomClaims(claims: unknown): asserts claims is Record<string,
   }
 }
 
+// generates a new random ID, checking against an optional set of "existing ids" for
+// uniqueness. if a unique ID cannot be generated in 10 tries, an internal error is
+// thrown. the ID generated by this method is not added to the set provided to this
+// method, callers must manage their own state.
+function newRandomId(length: number, existingIds?: Set<string>): string {
+  for (let i = 0; i < 10; i++) {
+    const id = randomId(length);
+    if (!existingIds?.has(id)) {
+      return id;
+    }
+  }
+  throw new InternalError(
+    "INTERNAL_ERROR : Failed to generate a random ID after 10 attempts",
+    "INTERNAL"
+  );
+}
+
+function getMfaEnrollmentsFromRequest(
+  state: ProjectState,
+  request: MfaEnrollments,
+  getEnrollmentId: (enrollment: MfaEnrollment, enrollmentIds: Set<string>) => string
```

Comment:
nit: I'd actually suggest just `options: {generateEnrollmentIds: boolean}` instead. Less code, less abstractions and more readable.